easy-ui-react/src/TextField/TextField.tsx

```diff
@@ -0,0 +1,359 @@
+    label,
+    errorText,
+    helperText,
+    placeholder,
+    value,
+    defaultValue,
+    iconAtStart,
+    iconAtEnd,
+  } = props;
+  const [isVisible, setIsVisible] = useState(false);
+  const ref = React.useRef(null);
+
+  const {
+    labelProps,
+    inputProps,
+    descriptionProps: helperTextProps,
+    errorMessageProps: errorTextProps,
+  } = useTextField(props, ref);
+
+  const bothIconPropsDefined = iconAtEnd && iconAtStart;
+  if (bothIconPropsDefined) {
+    // eslint-disable-next-line no-console
+    console.warn("Cannot simultaneously define `iconAtEnd` and `iconAtStart`");
+  }
+
+  const isPassword = type === "password";
+  const hasError = validationState === "invalid";
+  const showErrorText = hasError && errorText;
+  const showHelperText = !showErrorText && helperText;
+  const canUseIcon = !bothIconPropsDefined && !isPassword;
+  const hasStartIcon = canUseIcon && iconAtStart;
+  const hasEndIcon = canUseIcon && iconAtEnd;
+  const typeAdjustedForPasswordVisibility = isPassword && isVisible;
+
+  return (
+    <div className={classNames(styles.root)}>
+      <div className={styles.inputLabelContainer}>
+        <label
+          {...labelProps}
+          className={classNames(
+            styles.label,
+            isLabelVisuallyHidden && styles.labelHidden,
+          )}
+        >
+          {getLabelText(
+            emphasizedLabel,
+            hasError,
+            isLabelVisuallyHidden,
+            label,
+            size,
+          )}
+        </label>
+        <div className={styles.inputIconContainer}>
+          {hasStartIcon && getIcon(iconAtStart, true, size, isDisabled)}
+          <input
+            {...inputProps}
+            className={classNames(
+              styles.input,
+              isPassword && styles.passwordInput,
+              hasError && styles.errorInput,
+              hasStartIcon && styles.iconStartInput,
+              hasEndIcon && styles.iconEndInput,
+              styles[variationName("inputSize", size)],
+            )}
+            ref={ref}
+            type={typeAdjustedForPasswordVisibility ? "text" : type}
+            value={value}
+            required={isRequired}
+            disabled={isDisabled}
+            placeholder={placeholder}
+            autoFocus={autoFocus}
+            defaultValue={defaultValue}
+          />
+          {isPassword ? (
+            <UnstyledButton
+              className={classNames(
+                styles.passwordBtn,
+                hasError && styles.passwordBtnError,
+                styles[variationName("btnSize", size)],
+              )}
+              onPress={() => setIsVisible((prevVisibility) => !prevVisibility)}
+              isDisabled={isDisabled}
+            >
+              <Text visuallyHidden>password visibility</Text>
+              <Icon
+                symbol={isVisible ? VisibilityIcon : VisibilityOffIcon}
+                size={mapIconSize(size)}
+              />
+            </UnstyledButton>
```

Comment:
**question:** it may be overkill so up to you, but would it be worth creating a component for this to keep `TextField` main render body a little more streamlined? `<TextFieldPasswordButton />` or w/e

Reply:
I think its worth it, created a `PasswordButton` component